framework/source/class/qx/util/DynamicScriptLoader.js

```diff
@@ -0,0 +1,241 @@
+ *      src = '.src';
+ *      min = '';
+ *    }
+ *
+ *    // initialize the script loading
+ *    var dynLoader = new qx.util.DynamicScriptLoader();
+ *
+ *
+ *    dynLoader.addListenerOnce('ready',function(e){
+ *      console.log("all scripts have been loaded!");
+ *    });
+ *
+ *    dynLoader.addListener('failed',function(e){
+ *      var data = e.getData();
+ *      console.log("failed to load "+data.script);
+ *    });
+ *
+ *    dynLoader.load([
+ *        "myapp/jquery/jquery"+min+".js",
+ *        "myapp/highcharts/highcharts"+src+".js",
+ *        "myapp/highcharts/highcharts-more"+src+".js",
+ *        "myapp/highcharts/highcharts-modifications.js"
+ *    ]);
+ *    
+ * </code>
+ */
+qx.Class.define("qx.util.DynamicScriptLoader", {
+  extend: qx.core.Object,
+
+  /**
+   * Initialize the state hashes if not yet done
+   */
+
+  construct: function () {
+    this.base(arguments);
+    this.__QUEUE = [];
+    this.__LOADING = false;
+  },
+
+  /*
+  *****************************************************************************
+     EVENTS
+  *****************************************************************************
+  */
+
+  events: {
+    /**
+     * fired when a script is loaded successfully. The data contains 'script' and 'status' keys.
+     */
+    loaded: 'qx.event.type.Data',
+
+    /**
+     * fired when a specific script fails loading.  The data contains 'script' and 'status' keys.
+     */
+    failed: 'qx.event.type.Data',
+
+    /**
+     * fired when all given scripts are loaded, each time loadScriptsDynamic is called.
+     */
+    ready: 'qx.event.type.Event'
+  },
+
+  statics: {
+    /**
+     * Map of scripts being added at the present time. Key is script name; value is instance of this class which
+     * is loading it.
+     */
+    __IN_PROGRESS: {},
+     /**
+      * Map of scripts that have fully loaded. Key is script name; value is true
+      */
+    __LOADED: {}
+  },
+
+  /*
+  *****************************************************************************
+     MEMBERS
+  *****************************************************************************
+  */
+
+
+  members: {
+
+    __QUEUE: null,
+    __LOADING: null,
+
+
+    /**
+     * Load scripts dynamically, typically used in the instance constructor
+     *
+     * @param codeArr {Array} an array with the uri names of the scripts
+     *
+     */
+    load: function(codeArr) {
+      codeArr.forEach(function(script) {
+         this.__QUEUE.push(script);
+      },this);
+      this.__loadScripts();
+    },
+
+
+    /**
+     * Chain loading scripts.
+     *
+     * Recursively called until the array of scripts is consumed
+     *
+     * @param codeArr {Array} an array with the uri names of the scripts
+     */
+    __loadScripts: function() {
+      var cl = qx.util.DynamicScriptLoader;
+      var script;
+      var dynLoader;
+      var id1, id2;
+      var uri;
+      var loader;
+
+      if (this.__LOADING === true) {
+        return;
```

Comment:
The first call to load() causes __loadScripts to be called. When a script name to be loaded is found, that is not already loaded, __LOADING is set to true, and loading for that script (and, subsequently, any following not-previously-loaded ones on the list) is begun. If a new call to load() comes in on this same instance of DynamicScriptLoader, with additional scripts to be loaded, __loadScripts will be called and find __LOADING is already true, so those scripts which were added to the queue will be loaded when the first instance gets around to them. This new load() call can and does return immediately.

There is a bug, however. __LOADING is set to false in the "loaded" and "failed" handlers, whereas it should only be set to false in the "ready" handler (or rather, when there is nothing left on the queue, whether all scripts successfully loaded or not).

Again, though, if we got rid of the load() method and passed the list of scripts to be loaded directly to the constructor, this whole issue (and the __LOADING member) would disappear.